File: package/main/src/Math/primeFactorization.ts
/**
 * Performs prime factorization of a number
 * @param  {number} x Number to factorize
 * @returns {Array<{number: number; count: number}>} Array of prime factors and their counts
 * @example primeFactorization(12); // [{number: 2, count: 2}, {number: 3, count: 1}]
 * @description
 * Returns an array of objects containing prime factors and their counts.
 * For example, 12 = 2^2 * 3^1 is represented as [{number: 2, count: 2}, {number: 3, count: 1}]
 */
export const primeFactorization = (x: number) => {
  let n = 0;
  let copyX = x;
  const out: { number: number; count: number }[] = [];
  for (let index = 2; index * index <= copyX; index++) {
    if (copyX % index === 0) {
      n = 0;
      while (copyX % index === 0) {
        n++;
        copyX /= index;
      }
      out.push({ number: index, count: n });
    }
  }
  // If remaining value is greater than 1, it's a prime factor
  if (copyX > 1) {
    out.push({ number: copyX, count: 1 });
  }
  return out;
};

File: package/umt_python/src/math/prime_factorization.py
from typing import TypedDict


class PrimeFactor(TypedDict):
    number: int
    count: int


def prime_factorization(x: int) -> list[PrimeFactor]:
    """
    Performs prime factorization of a number.

    Args:
        x: Number to factorize.

    Returns:
        Array of prime factors and their counts.

    Example:
        >>> prime_factorization(12)
        [{'number': 2, 'count': 2}, {'number': 3, 'count': 1}]
    """
    if x <= 1:
        return []

    copy_x = x
    out: list[PrimeFactor] = []

    if copy_x % 2 == 0:
        n = 0
        while copy_x % 2 == 0:
            n += 1
            copy_x //= 2
        out.append({"number": 2, "count": n})

    index = 3
    while index * index <= copy_x:
        if copy_x % index == 0:
            n = 0
            while copy_x % index == 0:
                n += 1
                copy_x //= index
            out.append({"number": index, "count": n})
        index += 2
    # If remaining value is greater than 1, it's a prime factor
    if copy_x > 1:
        out.append({"number": copy_x, "count": 1})
    return out
